Enable live migration of VMs with hotplug volumes

vasiliy-ul opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

Currently live migation of VMs with attached hotplug volumes is not implemented and it gets rejected by validation hook. This PR adds the missing pieces to enable it.

**Special notes for your reviewer**:

The PR is based on https://github.com/kubevirt/kubevirt/pull/5649 (which refactors hotplug disks implementation to use a single attachment pod for all volumes). Also it relies on the fix https://github.com/kubevirt/kubevirt/pull/5436. The logic is implemented as follows:

- For the migration with the status `MigrationPending` after creating the `virt-launcher` target pod, `virt-controller` checks if the VMI has hotplug volumes attached. In case if it does a new target attachment pod is created (with the target `virt-launcher` pod set as owner).
- When the attachment pod is ready `virt-controller` updates the migration phase to `MigrationScheduling`.
- Once in `MigrationScheduled`, `virt-controller` adds a `MigrationTargetAttachmentPodLabel` to VMI with the UID of the attachment pod (the UID is needed for the `virt-handler` on the target node).
- During the preparation stage `virt-handler` on the target node takes the attachment pod UID and performs the bind mount of the hotplug volumes to the `virt-launcher` pod.

~Still in TODO~ Done:

- [x] merge the mentioned PRs 
- [x] cleanup the code
- [x] add test

**(!)** The PR needs a recent version of `runc` (at least `v1.0.0`) that includes the fix https://github.com/opencontainers/runc/commit/bf7492ee5d022cd99a9dbe71c5c4f965041552e9 . Otherwise the migration of VMs with block volumes to the nodes with CPU Manager and static policy may fail (the issue is raised and discussed in the mailing list: https://groups.google.com/g/kubevirt-dev/c/0UbMD2R92mQ)

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Live migration of VMs with hotplug volumes is now enabled
```

vasiliy-ul commented:
ping @awels , maybe you could take a look if the approach in this PR is viable?

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.17-rook-ceph

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.17-rook-ceph

awels commented on tests/storage/hotplug.go:
```diff
@@ -318,7 +318,7 @@ var _ = SIGDescribe("Hotplug", func() {
 	verifyCreateData := func(vmi *kubevirtv1.VirtualMachineInstance, device string) {
 		batch := []expect.Batcher{
-			&expect.BSnd{S: fmt.Sprintf("sudo mkfs.ext4 %s\n", device)},
+			&expect.BSnd{S: fmt.Sprintf("sudo mkfs.ext4 -F %s\n", device)},
```
This is a bad thing to do, -F forces a file system if one already exists, this usually means we are re-using a disk and its a bug.

vasiliy-ul replied:
Right, I will change that. Now its just a temporary fix for the new test with migration. The test starts a VM, attaches a volume and then does several migrations. Each time a migration is completed it checks whether the volume is still usable. I will refactor it to create the filesystem only once.

awels commented:
I see no particular issues with the approach. I just need to hurry up and finish #5649 I already fixed a few issues with it locally that I could push if you want to look at them.

vasiliy-ul commented:
> I see no particular issues with the approach. I just need to hurry up and finish #5649 I already fixed a few issues with it locally that I could push if you want to look at them.

Yeah, sure, that would be great. I tested the implementation with various storages: rook-ceph, longhorn, nfs. The flip-flop pod works nice and so does the migration on top of it. Though if you fixed locally several issues I guess I missed smth :) ... from my side I havent noticed serious problems so far.

awels commented:
Nothing I fixed was 'serious' but you found one in your PR, where you are why are we not returning an error. I am currently updating the unit tests and ensuring all the proper things are still being tested.

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.17-rook-ceph

awels commented:
/cc @davidvossel
Can you take a look at this PR as well. It implements LM with the single attachment pod. There is a current known issue with block devices which will fix itself with a newer version of runc.

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.20-sig-storage

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.19-sig-storage

vasiliy-ul commented:
/retest

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.20-sig-storage

vasiliy-ul commented:
/retest

vasiliy-ul commented:
/retest

vasiliy-ul commented:
There was an issue with scheduling the attachment pod after the rebase on the latest master. That caused failures in storage tests. I think now it should be fixed.

vasiliy-ul commented:
> /cc @davidvossel
> Can you take a look at this PR as well. It implements LM with the single attachment pod. There is a current known issue with block devices which will fix itself with a newer version of runc.

hey, ping @davidvossel , maybe you could have a quick look at the PR and give some feedback? Would be great if we could move this forward. Thanks.

weihanglo commented:
Hi @vasiliy-ul. After testing this PR manually in a 3-node cluster, I found that the migration only succeeds for the first time. Any subsequent migrations will fail with the following error, despite whether it has hotplug volumes or not:

```
{
  "component": "virt-launcher",
  "kind": "",
  "level": "error",
  "msg": "Live migration failed.",
  "name": "vm0",
  "namespace": "default",
  "pos": "live-migration-source.go:909",
  "reason": "error encountered during proxy setup: listen tcp [::1]:49153: bind: address already in use",
  "timestamp": "2021-07-06T07:29:58.508265Z",
  "uid": "b6554815-669b-411a-9068-171f1859d0be"
}
```

This error happened on the pod on the source node, which previously was also a target node during the first migration.

I doubt that we forgot to cleanup some migration proxies somewhere but after skimming the code I have no idea. In addition, this issue does not happen on v0.40.0, v0.42.1 and v0.43.0-rc.0.

vasiliy-ul commented:
Hi @weihanglo . Hm... I am able to reprodue the issue only with our downstrream build. When using kubevirt build from the master branch everything works fine. But anywa, thanks for reporting this. I will take a look.

vasiliy-ul commented:
I think the error happens because the golang `net` package does not set `SO_REUSEPORT` for `net.Listener` and the migration proxy uses a predefined set of port for migration:

https://github.com/kubevirt/kubevirt/blob/36496f394c77099ec950f178f7669e1b91e59c23/pkg/virt-handler/migration-proxy/migration-proxy.go#L39-L42

https://github.com/kubevirt/kubevirt/blob/36496f394c77099ec950f178f7669e1b91e59c23/pkg/virt-handler/migration-proxy/migration-proxy.go#L384-L388

I also checked without hotplugged disks and the issue is still reproducable. So I guess its not related to the changes in this PR. The only thing I am wondering now is why it does not happed with the original images. For now my only assumption is that some OS tunables may affect the behaviour.

vasiliy-ul commented:
> I think the error happens because the golang net package does not set SO_REUSEADDR for net.Listener and the migration proxy uses a predefined set of ports for migration:

After some investigation looks like my assumption was wrong. `SO_REUSEADDR` does not seem to solve the issue. Though anyway the failure happens only with our custom images and it is related to migration in general (it is reproducable even without the hotplug volumes). I will try to debug it further. Meanwhile I think this PR is still ready for review.

davidvossel commented:
> hey, ping @davidvossel , maybe you could have a quick look at the PR and give some feedback? Would be great if we could move this forward. Thanks.

yep no problem

davidvossel commented on pkg/virt-controller/watch/migration.go:
```diff
@@ -668,6 +699,62 @@ func (c *MigrationController) handleTargetPodCreation(key string, migration *vir
 }
 
+func (c *MigrationController) createAttachmentPod(migration *virtv1.VirtualMachineInstanceMigration, vmi *virtv1.VirtualMachineInstance, virtLauncherPod *k8sv1.Pod) error {
+	sourcePod, err := controller.CurrentVMIPod(vmi, c.podInformer)
+	if err != nil {
+		return fmt.Errorf("failed to get current VMI pod: %v", err)
+	}
+
+	volumes := getHotplugVolumes(vmi, sourcePod)
+	volumeNamesPVCMap := make(map[string]*k8sv1.PersistentVolumeClaim)
+	for _, volume := range volumes {
+		claimName := kubevirttypes.PVCNameFromVirtVolume(volume)
+		if claimName == "" {
+			log.Log.Object(migration).Infof("Unable to hotplug, volume %s not PVC or Datavolume", volume.Name)
+			continue
+		}
+		pvc, exists, _, err := kubevirttypes.IsPVCBlockFromStore(c.pvcInformer.GetStore(), virtLauncherPod.Namespace, claimName)
+		if err != nil {
+			return fmt.Errorf("failed to get pvc: %v", err)
+		}
+		if !exists {
+			log.Log.Object(migration).Infof("Unable to hotplug, claim %s not found", claimName)
+			continue
+		}
+		volumeNamesPVCMap[volume.Name] = pvc
+	}
```
since this is duplicate logic in vmi.go, can we get a shared util function for this. I know it will require a bit refactoring.

vasiliy-ul replied:
sure. will think how to refactor it.

davidvossel commented on pkg/virt-controller/watch/migration.go:
```diff
@@ -399,7 +403,21 @@ func (c *MigrationController) updateStatus(migration *virtv1.VirtualMachineInsta
+					}
+					if len(attachmentPods) > 0 {
+						attachmentPod := attachmentPods[0]
+						if isPodReady(attachmentPod) {
```
I'd rather the pod "ready" condition for both the target and attachment pod be an indication that the migration is "scheduled"

Here we should just be checking that the attachment pod exists.

davidvossel commented on pkg/virt-controller/watch/migration.go:
```diff
@@ -576,6 +594,19 @@ func (c *MigrationController) handleTargetPodHandoff(migration *virtv1.VirtualMa
+		}
+		if len(attachmentPods) > 0 {
+			log.Log.Object(migration).Infof("Target attachment pod for vmi %s/%s: %s", vmiCopy.Namespace, vmiCopy.Name, string(attachmentPods[0].UID))
+			vmiCopy.ObjectMeta.Labels[virtv1.MigrationTargetAttachmentPodLabel] = string(attachmentPods[0].UID)
```
Is there a reason this needs to be a label? I'd think this would be something we'd store on the vmi.Status.MigrationState struct. 

we use a label for the NodeName because we have an informer that is using that label as a selector

vasiliy-ul replied:
No specific reason. Right, probably the struct would be a better place.

davidvossel commented on tests/storage/hotplug.go:
```diff
@@ -833,10 +851,15 @@ var _ = SIGDescribe("Hotplug", func() {
 		})
 
-		Context("VMI only", func() {
+		Context("[Serial]VMI migration", func() {
```
does this need to be serial?

vasiliy-ul replied:
The test relies on the workaround for the issue with CPU manager and runc: there is a need to ensure that VMI will be migrated to the node **without** cpu manager. For that I do the manipulations with the nodes and set them schedulable/unschedulable as needed. If I understand correctly, the `Serial` label here will prevent other tests to run in parallel with this one. Potentially such nodes manipulations may affect other tests I guess. Or do you think it is not needed in this case?

davidvossel replied:
ohh interesting. i think i missed that the node stuff was going on.

There's another way to do this that's super simple and probably doesn't require setting nodes as unscheduable.

- Pick a single node out that you want to run the vmi on and add a unique label to the node
- set a node selector on the VMI to ensure it get's scheduled to that node
- once the VMI is online, add that label to the node you want the VMI to migrate to
- the migration's target pod will then choose that node to run on

davidvossel reviewed (commented):
This complicates things a bit with migrations, but i think it's a viable approach.

I made a few comments in line, and here are some more broad comments.

- We need a validation webhook to prevent add/remove of volumes when the VMI is being migrated.
- We never want to get into a state where the migration can never complete (whether successfully or unsuccessfully), so i think we need to make sure the migration fails if the attachment pod reaches a finalized state. Similar to this logic in updateStatus.

```
        } else if podExists && podIsDown(pod) {
                migrationCopy.Status.Phase = virtv1.MigrationFailed
                c.recorder.Eventf(migration, k8sv1.EventTypeWarning, FailedMigrationReason, "Migration failed because target pod shutdown during migration")
                log.Log.Object(migration).Errorf("target pod %s/%s shutdown during migration", pod.Namespace, pod.Name)
```

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.20-sig-storage

vasiliy-ul commented:
hey @davidvossel , thank you for review. I think I've addressed all the comments. Could you maybe take one more look? I am probably not absolutely sure about the validation webhook though, i.e. if I put it in the right place.

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.20-sig-storage

davidvossel commented on pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:
```diff
@@ -438,6 +438,14 @@ func (admitter *VMsAdmitter) validateVolumeRequests(vm *v1.VirtualMachine) ([]me
 		if len(causes) > 0 {
 			return causes, nil
 		}
+
+		if vmi.Status.MigrationState != nil && !vmi.Status.MigrationState.Completed {
+			return []metav1.StatusCause{{
+				Type:    metav1.CauseTypeFieldValueNotSupported,
+				Message: fmt.Sprintf("Cannot handle volume requests while VMI migration is in progress"),
+				Field:   k8sfield.NewPath("spec").String(),
+			}}, nil
+		}
```
for this condition, can you use the `pkg/util/migrations/migrations.go` file's IsMigrating function to determine if a migration is in flight?

I'm not 100% sure Completed is always an accurate indicator.

vasiliy-ul replied:
NP, done. Replaced the condition with `IsMigrating`.

davidvossel reviewed (commented):
/approve

looks great, I'm giving a preemptive approve. I just had one super minor comment

vasiliy-ul commented:
/retest

vasiliy-ul commented:
hey, since all the comments and issues are addressed now and the PR is approved, maybe you guys could also lgtm it? @davidvossel , @awels what do you think?

awels reviewed (approved):
/lgtm

awels commented:
/test pull-kubevirt-e2e-k8s-1.19-sig-compute
